src/include/index/flatpq_index.h

```diff
@@ -0,0 +1,927 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Header-only library of class that implements a flat index that used product
+ * quantization.
+ *
+ */
+
+#ifndef TILEDB_FLATPQ_INDEX_H
+#define TILEDB_FLATPQ_INDEX_H
+
+#include <cstddef>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <unordered_set>
+
+#include "detail/flat/qv.h"
+#include "detail/linalg/matrix.h"
+#include "detail/linalg/tdb_io.h"
+
+#include <tiledb/group_experimental.h>
+#include <tiledb/tiledb>
+
+template <class T, class U>
+void sub_kmeans_random_init(
+    const ColMajorMatrix<T>& training_set,
+    ColMajorMatrix<U>& centroids,
+    size_t sub_begin,
+    size_t sub_end,
+    size_t seed = 0) {
+  scoped_timer _{__FUNCTION__};
+
+  if (num_vectors(training_set) < num_vectors(centroids)) {
+    throw std::invalid_argument(
+        "Number of vectors in training set must be greater than or equal to "
+        "number of centroids");
+  }
+
+  std::mt19937 gen(seed == 0 ? std::random_device{}() : seed);
+  std::uniform_int_distribution<> dis(0, num_vectors(training_set) - 1);
+
+  size_t num_clusters = num_vectors(centroids);
+
+  std::vector<size_t> indices(num_clusters);
+  std::unordered_set<size_t> visited;
+  for (size_t i = 0; i < num_clusters; ++i) {
+    size_t index;
+    do {
+      index = dis(gen);
+    } while (visited.contains(index));
+    indices[i] = index;
+    visited.insert(index);
+  }
+  // std::iota(begin(indices), end(indices), 0);
+  // std::shuffle(begin(indices), end(indices), gen);
+
+  for (size_t i = 0; i < num_clusters; ++i) {
+    for (size_t j = sub_begin; j < sub_end; ++j) {
+      centroids(j, i) = training_set(j, indices[i]);
+    }
+  }
+}
+
+// #define REASSIGN
+
+/**
+ * @brief Run kmeans on a subspace
+ * @tparam T Data type of the input vectors
+ * @param training_set Training set
+ * @param centroids Initial locations of centroids.  Will be updated in
+ * locations [sub_begin, sub_end).
+ * @param sub_begin Beginning of the subspace
+ * @param sub_end End of the subspace
+ * @param num_clusters Number of clusters to find
+ * @return
+ *
+ * @todo update with concepts
+ * @todo fix up zero sized partitions
+ * @todo this would be more cache friendly to do sub_begin and sub_end
+ * in an inner loop
+ * @todo We can probably just reuse plain kmeans at some point.
+ */
+template <class T, class U>
+auto sub_kmeans(
```

Comment:
Can we use the existing kmeans library for this task?

Reply:
We should definitely integrate them as much as we can.  The main difference between the two is whether or not to operate on a subset of the vectors or on the whole vector (there are some small differences as well).  Since the whole vector is just a particular subset, the same code should be usable.  I have created a story for this.